### ISSUE TITLE
Proposal: Consolidate SkyAtmosphere and Globe atmosphere settings?

### ISSUE THREAD

**Issue by ptrgags** [CesiumGS/cesium#11681]
As I'm looking into #4196, one thing I notice is that the following settings exist on both `viewer.scene.skyAtmosphere` (for the sky) and `viewer.scene.globe` (for ground and fog for terrain):

* `atmosphereBrightnessShift` / `brightnessShift`
* `atmosphereHueShift` / `hueShift`
* `atmosphereLightIntensity`
* `atmosphereMieAnisotropy`
* `atmosphereMieCoefficient`
* `atmosphereMieScaleHeight`
* `atmosphereRayleighCoefficient`
* `atmosphereRayleighScaleHeight`
* `atmosphereSaturationShift` / `saturationShift`

Ultimately these impact the shader uniforms in `AtmosphereCommon.glsl`:

```
uniform vec3 u_radiiAndDynamicAtmosphereColor;
uniform float u_atmosphereLightIntensity;
uniform float u_atmosphereRayleighScaleHeight;
uniform float u_atmosphereMieScaleHeight;
uniform float u_atmosphereMieAnisotropy;
uniform vec3 u_atmosphereRayleighCoefficient;
uniform vec3 u_atmosphereMieCoefficient;
```

Since the ground and sky atmosphere both represent the Earth's atmosphere, conceptually they should always use the same settings.

There's also two other benefits of combining them:

1. This streamlines the interface. Instead of changing settings in two places (skyAtmosphere and globe), you only need to change one (I'm thinking `Globe`, but open to suggestions).
2. It also would streamline things deeper in the implementation details. In #4196 I'm already considering changing the Globe's implementation to use `AutomaticUniforms`, this way it's easier to implement fog for 3D Tiles.

I'm curious if the community would find this to be a reasonable breaking change, or if there are use cases where configuring `SkyAtmosphere` and `GroundAtmosphere` differently is helpful. I'll ask about this on the forum as well.

**Comment by lilleyse** [comment]
> I'm thinking `Globe`, but open to suggestions

I think the atmosphere options should be in its own object under `Scene` in case you don't have a globe

**Comment by ptrgags** [comment]
@lilleyse That's a good point, I like that for 2 more reasons:

1. It puts the parameters in their own dedicated struct to make 
2. The implementation would have more symmetry with how `scene.fog` is implemented

**Comment by ptrgags** [comment]
I made a [forum post](https://community.cesium.com/t/proposed-breaking-change-consolidate-atmosphere-settings-for-skyatmosphere-and-globe/28591) to see if the community has any further thoughts on this.

**Comment by ptrgags** [comment]
A couple more thoughts:

1. For non-photorealistic rendering, it's kinda neat to be able to adjust the color of sky atmosphere and ground atmosphere/fog separately. So maybe the HSB shift parameters could be duplicated for each system, but keep the other atmosphere parameters in sync.
![image](https://github.com/CesiumGS/cesium/assets/8422414/77b9c2a7-201e-486a-8784-87e2bb1fec95)
2. The flags that enable lighting for the sky/ground/fog are rather tangled. Here's a table to keep track of what affects what.

| Setting | Purpose | (`Globe`) ground atmosphere | `SkyAtmosphere` | (`Globe`) Terrain fog | Fog culling |
|---|---|---|---|---|--|
| `Viewer` constructor: `options.globe` | If false, the globe is not created/rendered | Ground atmosphere is not rendered.. `SkyAtmosphere` is not directly affected, but will not use lighting from sun. Fog is not rendered. |
| `globe.enableLighting` | turns on lighting from the sun or other light source | Toggles lighting | toggles lighting (flag synced via `Scene`) | Toggles lighting regardless of `globe.showGroundAtmosphere` | No effect |
| `globe.dynamicAtmosphereLighting` | Use the sun or other light source. If false, use a light directly overhead of the vertex/fragment | Toggles dynamic lighting | Toggles dynamic lighting (flag synced via `Scene`) | Toggles dynamic lighting (regardless of `globe.showGroundAtmosphere` | No effect |
| `globe.dynamicAtmosphereLightingFromSun` | forces using the sun rather than the scene's current light source (sun or directional light) | Toggles forced sunlight | Toggles forced sunlight (flag synced via `Scene`) | Toggles forced sunlight regardless of `globe.showGroundAtmosphere` | No effect |
| `globe.showGroundAtmosphere` | Turns off rendering of the ground atmosphere | Not rendered | No effect | No effect | No effect |
| `skyAtmosphere.show` | Turns off rendering the sky atmosphere | No effect | Not rendered | No effect | No effect |
| `fog.enabled` | Turn off anything to do with fog. | No effect | no effect | Not rendered | No fog culling |
| `fog.renderable` | Toggle showing fog, but still allow fog culling | No effect | No effect | Not rendered | No effect |

Some of these settings might be better off stored in the new `scene.atmosphere` so all the systems always are synced to it via `FrameState` and/or `AutomaticUniforms`

**Comment by ptrgags** [comment]
After talking with @ggetz the other day, we think that the above is a bit of overkill. A couple changes to streamline the above:

1. a separate `GroundAtmosphere` class isn't really needed, just add the settings to `scene.atmosphere` (except for `show` which should be scoped to the primitive responsible for rendering, whether that's the globe or the tileset)
2. For consistency and ease of use, just have a single place for hue/saturation/brightness shift (also in `scene.atmosphere`). This means the colors for sky and ground atmosphere will always be in sync (before they were independent.

So to summarize the overall architecture

* `scene.atmosphere` will contain atmosphere settings for any of the systems that need it and update the `FrameState`
* `scene.skyAtmosphere` will still be there because it's needed for rendering the sky, but some of the parameters will move to `scene.atmosphere`
* `scene.globe` will have some of its parameters moved to `scene.atmosphere`

**Comment by ptrgags** [comment]
Some implementation details from #11744 that will need changes after we change the API:

1. `FogStage(FS|VS).glsl` now use built-in functions for scattering such as `czm_computeGroundAtmosphereScattering()`, `czm_getDynamicAtmosphereLightDirection()`, etc. These ones use automatic uniforms, compared with the old globe shader that supplied the uniforms itself. Once the API changes to move parameters out of globe, switch to these new functions and remove the dead code that remains
2. in `Scene#updateEnvironment()`, there's a section of code that calls `skyAtmosphere.setDynamicLighting()` from the globe flags. The `if defined(globe)` branch will no longer be needed once all atmosphere settings come from `scene.atmosphere`, the else clause will always run.

**Comment by ptrgags** [comment]
Some thoughts about how to make the `Globe` and `SkyAtmosphere` settings work during the deprecation period:

1. Whenever `Scene.globe`'s setter is called, also set `globe._atmosphere = this.atmosphere` so the globe now has a reference to the atmosphere struct.
2. Change `Scene.skyAtmosphere` to a setter, and do a similar `skyAtmosphere._atmosphere = this.atmosphere`
3. Make a `Scene.atmosphere` setter that updates `globe._atmosphere` and `skyAtmosphere._atmosphere` to the new atmosphere object (if the globe/sky exist)
4. For `Globe`/`SkyAtmosphere`, turn all of the atmosphere-related members into properties, and the setters should: A. log a deprecation warning, and B. Don't update the globe/sky, update `_atmosphere.<propertyNameHere>` instead (checking first if `_atmosphere` is defined). If it's not defined... maybe another warning is needed (not the most elegant, but this is temporary for the deprecation period)
5. In the shaders, use the `czm_atmosphere*` builtins and functions
6. Also remove all old atmosphere uniforms manually set by the globe/sky atmosphere (`u_atmosphere*`) and the code used to configure them, as these will become dead code (note, make sure to check through `GlobeSurfaceTileProvider.js` and `GlobeSurfaceShaderSet.js` carefully.)
